lib/logstash/outputs/http.rb

```diff
@@ -55,88 +60,155 @@ class LogStash::Outputs::Http < LogStash::Outputs::Base
-    when "post"
-      request = @agent.post(event.sprintf(@url))
+    request.on_failure do |exception|
+      log_failure("Could not fetch URL",
+                  :url => url,
+                  :method => @http_method,
+                  :body => body,
+                  :headers => headers,
+                  :message => exception.message,
+                  :class => exception.class.name,
+                  :backtrace => exception.backtrace
+      )
+    end
+  end
+
+  private
+
+  # This is split into a separate method mostly to help testing
+  def log_failure(message, opts)
+    #puts "FAILED #{message} #{opts}"
+    @logger.error("[HTTP Output Failure] #{message}", opts)
+  end
+
+  # Manticore doesn't provide a way to attach handlers to background or async requests well
+  # It wants you to use futures. The #async method kinda works but expects single thread batches
+  # and background only returns futures.
+  # Proposed fix to manticore here: https://github.com/cheald/manticore/issues/32
+  def request_async_background(request)
+    @method ||= client.executor.java_method(:submit, [java.util.concurrent.Callable.java_class])
+    @method.call(request)
+  end
+
+  # Format the HTTP body
+  def event_body(event)
+    # TODO: Create an HTTP post data codec, use that here
+    if @format == "json"
+      LogStash::Json.dump(map_event(event))
+    elsif @format == "message"
+      event.sprintf(@message)
     else
-      @logger.error("Unknown verb:", :verb => @http_method)
+      encode(map_event(event))
     end
+  end
 
-    if @headers
-      @headers.each do |k,v|
-        request.headers[k] = event.sprintf(v)
+  def map_event(event)
+    if @mapping
+      @mapping.reduce({}) do |acc,kv|
+        k,v = kv
+        acc[k] = event.sprintf(v)
+        acc
       end
+    else
+      event.to_hash
     end
+  end
 
-    request["Content-Type"] = @content_type
+  def event_headers(event)
+    headers = custom_headers(event) || {}
+    headers["Content-Type"] = @content_type
+    headers
+  end
 
-    begin
-      if @format == "json"
-        request.body = LogStash::Json.dump(evt)
-      elsif @format == "message"
-        request.body = event.sprintf(@message)
-      else
-        request.body = encode(evt)
-      end
-      #puts "#{request.port} / #{request.protocol}"
-      #puts request
-      #puts
-      #puts request.body
-      response = @agent.execute(request)
-
-      # Consume body to let this connection be reused
-      rbody = ""
-      response.read_body { |c| rbody << c }
-      #puts rbody
-    rescue Exception => e
-      @logger.warn("Unhandled exception", :request => request, :response => response, :exception => e, :stacktrace => e.backtrace)
+  def custom_headers(event)
+    return nil unless @headers
+
+    @headers.reduce({}) do |acc,kv|
+      k,v = kv
+      acc[k] = event.sprintf(v)
+      acc
     end
-  end # def receive
+  end
 
+  #TODO Extract this to a codec
```

Comment:
I spoke to @suyograo about this earlier, we decided to triage this guy. I'd like to time box this and move on to the rabbitmq-input refactor which has some more serious issues.